Issue 41 - Raise Property Change on Expression Bodied Properties

Syrus666 opened the pull request:
Updated the ReactiveUIPropertyWeaver to support expression bodied properties.
This is achieved by adding this.RaisePropertyChanged("ExpressionBodiedPropertyName") to the setter on dependent properties decorated with the [Reactive] attribute, dependent properties are calculated recursively, so given

```C#
[Reactive] public string Foo { get; set; }
[Reactive] public string Bar { get; set; }
public string Baz { get; set; }

[Reactive] public string FooBar => Foo + Bar;
[Reactive] public string FooBarHello => FooBar + "hello";
[Reactive] public string FooBarHelloBaz => FooBarHello + Baz;
[Reactive] public string BazHello => Baz + "hello";
```

**FooBar** will raise on **Foo** or **Bar** being set.
**FooBarHello** will raise on **Foo** or **Bar** being set
**FooBarHelloBaz** will raise on **Foo** or **Bar** being set, but not Baz as it's not decorated with the [Reactive] attribute
**BazHello** will throw an **error during build** as non of the dependent properties in the chain are decorated with the [Reactive] attribute.

Syrus666 commented:
https://github.com/kswoll/ReactiveUI.Fody/issues/41

extesy commented:
@Syrus666 Would `FooBar` still work if you also mark `Foo` and `Bar` as `[Reactive]`?

Syrus666 commented:
@extesy Yup - tested here: https://github.com/kswoll/ReactiveUI.Fody/pull/42/files#diff-a29a00a1dd3a77ad1d12ad6dee83e8d9R31 (as [StringProperty](https://github.com/kswoll/ReactiveUI.Fody/pull/42/files#diff-a29a00a1dd3a77ad1d12ad6dee83e8d9R91) is marked with [Reactive])

Syrus666 commented on ReactiveUI.Fody/ReactiveUIPropertyWeaver.cs:
```diff
@@ -42,10 +42,14 @@ public void Execute()
 
+            var raisePropertyChangedMethod = ModuleDefinition.ImportReference(reactiveObjectExtensions.Methods.Single(x => x.Name == "RaisePropertyChanged"));
+            if (raisePropertyChangedMethod == null)
+                throw new Exception("raiseAndSetIfChangedMethod is null");
```
Ooops, copypasta fail.

kswoll commented on ReactiveUI.Fody/CecilExtensions.cs:
```diff
@@ -126,6 +126,15 @@ public static bool CompareTo(this TypeReference type, TypeReference compareTo)
             return type.FullName == compareTo.FullName;
         }
 
+        public static bool IsExpressionBodiedProperty(this MethodDefinition method)
```
I think this name is a bit confusing?  There should be no functional difference between an expression-bodied method:

    [Reactive]
    public string FooBar => Foo + Bar;

And a normal get-only method:

    [Reactive]
    public string FooBar
    {
        get { return Foo + Bar; }
    }

Right?  Perhaps `IsDependentProperty` would make more sense?

kswoll commented on ReactiveUI.Fody/ReactiveUIPropertyWeaver.cs:
```diff
@@ -54,6 +58,59 @@ public void Execute()
+                        genericTargetType = genericDeclaration;
+                    }
+
+                    if (property.SetMethod == null && property.GetMethod.IsExpressionBodiedProperty())
```
I think it might be slightly more performant and be a little more clear if `IsExpressionBodiedProperty` were renamed to `GetMethodDependencies` and simply return a `MethodDefinitinon[]`  Then you could just store that (perhaps in a variable named `methodDependencies`), check `.Any()` on it, and not need to iterate over the instructions again a few lines below.

And since you'd be starting from a set of `MethodDefinition`, the code for `setMethodsForGetInstructions` would read a little better, and you could probably just do `set.Contains(prop.GetMethod)`.   Going further, I think it might make sense to store the getter/setter mapping in a dictionary outside of the loop:

```
var setMethodsByGetMethod = targetType.Properties
    .Where(x => x.SetMethod != null && x.GetMethod != null)
    .ToDictionary(x => x.GetMethod);
```

That way we avoid getting into O(N^2) territory on these loops/linq operations.

kswoll commented on ReactiveUI.Fody/ReactiveUIPropertyWeaver.cs:
```diff
@@ -54,6 +58,59 @@ public void Execute()
+                                        prop.SetMethod.HasBody &&
+                                        getInstructions.Any(
+                                            i => ((MethodDefinition) i.Operand).FullName == prop.GetMethod.FullName))
+                                .Select(prop => prop.SetMethod);
```
Per the above comment, this could then be simplified to:

```
var setMethodsForGetInstructions = methodDependencies
    .Where(x => setMethodsByGetMethod.ContainsKey(x => x))
    .Select(x => setMethodsByGetMethod[x])
    .ToArray();
```

(The `.ToArray()` to avoid multiple traversals of the enumerable)

Syrus666 replied:
On small enumerables (like the number of set & get instructions I'd expect in this area) it can be more performant to do the multiple traversals than to allocate and populate an array. 
:+1: to the change though as it's not worth spinning up dotTrace for this 😄 
Would you rather ToList() though as it should allocate one less address space?

kswoll replied:
Huh, I have no strong preference, but I'm curious why `.ToList()` would allocate a little less space?

Syrus666 replied:
I was referring to this http://stackoverflow.com/a/16323412

kswoll commented on ReactiveUI.Fody/ReactiveUIPropertyWeaver.cs:
```diff
@@ -54,6 +58,59 @@ public void Execute()
+                        {
+                            method.Body.Emit(il =>
+                            {
+                                var last = method.Body.Instructions.Last(i => i.OpCode == OpCodes.Ret && (i.Next == null || method.Body.Instructions.Last() == i));
```
I believe all methods (even void methods) must have an `OpCodes.Ret`, so probably the secondary check is not necessary?  However, I wonder what would happen if you had a property set method implemented like:

```
public string Foo
{
    get => foo;
    set {
        if (bar)
        {
            this.foo = value;
            return;
        }
        this.foo = "baz";
    }
}
```

I think your PR will fail to propagate the `RaisePropertyChange` when `bar` is `true` since this method contains two returns and you're only adding your logic to the last one.  A little annoying, but it'd probably make sense to get all the return instructions and instrument the below three lines in front of each.

Syrus666 replied:
Great catch and you are correct.
It got me thinking though, what about scenarios such as
`if(1 == value) return;`
In this scenario we wouldn't want to raise a property change event. So we could either take the hit, or I can have a think of a different way to solve this problem.

Syrus666 replied:
I could also restrict so that it only works with dependent auto properties with the `[Reactive]` attribute to remove some of the edge cases, risk of it not behaving as expected with other weavers then though..?

kswoll replied:
Hmm, good point.  I think your second idea feels a little cleaner?  Like, just explicitly opt-in with `[Reactive]` or `[ObservableAsProperty]`?  Dunno, just considering the ramifications.

kswoll commented on ReactiveUI.Fody.Tests/Issues/Issue41Tests.cs:
```diff
@@ -0,0 +1,107 @@
+        {
+            // Arrange
+            var model = new TestModel();
+            var expectedInpcPropertyName = "DerivedProperty";
```
nit: change all these strings to use the `nameof` operator:  i.e. `nameof(TestModel.DerivedProperty)`

Syrus666 replied:
:facepalm:

Syrus666 commented:
Thanks for the speedy review @kswoll I'll try and address during some downtime this week.
Would you prefer a commit addressing the review feedback, then a squash post thumbs or the changes to go directly in to the current commit?

kswoll commented:
@Syrus666 my pleasure, and equal thanks to you to implementing this feature.  I think it's great!  But as for your question, I figure don't worry about squashing since GitHub will let us squash before the merge?

Syrus666 commented:
I've deviated from the review feedback slightly (see new commit) in that this works recursively to find dependent properties within the class. 
Check out the updated description above for details (FooBarHello and FooBarHelloBaz) 😄

Syrus666 commented:
Hmm... it seems AppVeyor doesn't like my C# 7.0 syntax [here](https://github.com/kswoll/ReactiveUI.Fody/pull/42/commits/1bf0ce22fbabde909e4b1bbb9f746302d6ae7b0f#diff-0a58e0bc6d90facfeb14358329f54da2R76)

Can this be resolved with AppVeyor or do you need me to fix in the PR?

kswoll commented:
Ah, interesting.  Yeah, need to update the appveyor build script.  I'll get to that today.  Once that's done, this looks good, so will merge it in after.  Thanks again!
